### DISCUSSION TITLE
How to share generated clients as package that allows passing of requestconfig

### DISCUSSION THREAD

**Discussion by christian-draeger** [orval-labs/orval#816] [Q&A] [answered]
Scenario:
I have multiple applications that uses the exact same api. I already generate the react-query clients on top of a custom axios instance as part of these applications/projects. 
Every application sets different request headers and a different axios baseurl, beside that the orval setup is an exact copy per application. I archived that by providing a mutator. 

Now I created a package/Lib that contains the  generated clients client methods and types to be able to maintain it in one place. 

When using the react-query hooks provided by my package in the applications I need to somehow set axios base url and request headers (authorization header) based on logic that lives inside the applications.

How can I pass axios baseurl and request headers to an already generated client?

**Comment by NimmLor** [comment, marked as answer]
Hi, I'm having a similar usecase where I need to set a header dynamically. I'm just importing the custom axios instance that is used by orval and setting the default values.

```tsx
// apps/webapp/src/App.tsx

import { AXIOS_INSTANCE } from '@packages/api/axiosInstance'

AXIOS_INSTANCE.defaults.headers.common = {
  ...AXIOS_INSTANCE.defaults.headers.common,
  'x-tenant': tenantId,
}

// you could set the baseURL like this:
AXIOS_INSTANCE.defaults.baseURL = `https://example.com`

const App = () => {
  return <div>Hello World</div>
}

export default App
```

```ts
// packages/api/axiosInstance.ts

import { getResolvedDomains } from '@packages/constants'
import axios, { type AxiosError, type AxiosRequestConfig } from 'axios'

export const AXIOS_INSTANCE = axios.create({
  baseURL: `https://${getResolvedDomains().apiRecord}`,
})

export const axiosInstance = async <T>(
  config: AxiosRequestConfig
): Promise<T> => {
  // eslint-disable-next-line import/no-named-as-default-member
  const source = axios.CancelToken.source()

  const promise = AXIOS_INSTANCE({
    ...config,
    baseURL: `https://${getResolvedDomains().apiRecord}`,
    cancelToken: source.token,
  }).then(({ data }) => data)

  // @ts-expect-error: as per orval example
  promise.cancel = () => {
    source.cancel('Query was cancelled by Vue Query')
  }

  return await promise
}

export default axiosInstance
```
